Stride-Labs/stride
# add inner bounds

**summary**
add inner bounds

**tests**
- unittests
- dockernet
    - verify bounds are set upon registration
    - verify hz halts if the inner bound is crossed

```
➜  stride git:(tight-bounds) ✗ stridedl q stakeibc show-host-zone GAIA       
host_zone:
...
  last_redemption_rate: "1.000000000000000000"
  max_inner_redemption_rate: "1.500000000000000000"
  max_redemption_rate: "1.500000000000000000"
  min_inner_redemption_rate: "0.900000000000000000"
  min_redemption_rate: "0.900000000000000000"
  redemption_rate: "1.000000000000000000"
```

update bounds
```
$STRIDE_MAIN_CMD tx stakeibc set-redemption-rate-bounds GAIA 1 1.4 --from $STRIDE_ADMIN_ACCT -y | TRIM_TX
```
and query them
```
  max_inner_redemption_rate: "1.400000000000000000"
  max_redemption_rate: "1.500000000000000000"
  min_inner_redemption_rate: "1.000000000000000000"
  min_redemption_rate: "0.900000000000000000"
```

try setting from the wrong account
```
$STRIDE_MAIN_CMD tx stakeibc set-redemption-rate-bounds GAIA 1.1 1.3 --from val1 -y | TRIM_TX
```
and the error
```
➜  stride git:(tight-bounds) ✗ bash dockernet/scripts/set_bound.sh
Error: invalid creator address (stride1uk4ze0x4nvh4fk0xm4jdud58eqn4yxhrt52vv7): invalid address
Usage:
  strided tx stakeibc set-redemption-rate-bounds [chainid] [min-bound] [max-bound] [flags]

Flags:
```

Set tight bound and observe halt
```
$STRIDE_MAIN_CMD tx stakeibc set-redemption-rate-bounds GAIA 1 1.000001 --from $STRIDE_ADMIN_ACCT -y | TRIM_TX
```

halt
```
dockernet-stride1-1  | 2:11AM ERR IsRedemptionRateWithinSafetyBounds check failed 1.001028241739306108 is outside inner safety bounds [1.000000000000000000, 1.000001000000000000] module=x/stakeibc
dockernet-stride1-1  | 2:11AM ERR [INVARIANT BROKEN!!!] GAIA's RR is 1.001028241739306108. ERR: IsRedemptionRateWithinSafetyBounds check failed 1.001028241739306108 is outside inner safety bounds [1.000000000000000000, 1.000001000000000000]: redemption rate outside safety bounds module=x/stakeibc

```
